### PR TITLE
Fix floodgate allocation in C++/Lua and well and pressure plate allocation in Ruby.

### PR DESCRIPTION
**Pull request by BenLubar**

### CODE REVIEW

File: library/modules/Buildings.cpp [hunk 1]
```diff
@@ -369,6 +370,12 @@ df::building *Buildings::allocInstance(df::coord pos, df::building_type type, in
                 obj->ready_timeout = 500;
             break;
         }
+    case building_type::Floodgate:
+        {
+            auto obj = (df::building_floodgatest*)bld;
```
Comment (lethosor) [inline library/modules/Buildings.cpp]: virtual_cast would be better here.

Comment (BenLubar) [inline library/modules/Buildings.cpp]: Should I change the other cases as well?

**Comment by lethosor** [comment]
Turns out this was the issue that kept floodgates from working in buildingplan as well.

File: library/modules/Buildings.cpp (hunk 1 above)
Comment (warmist) [inline library/modules/Buildings.cpp]: Isn't the static_cast correct way of casting in this case?

**Pull request merged**